### PR TITLE
[BEAM-5650]: Modify BoundedToUnboundedSourceAdapter to configure its reader to read more than 1 bounded source

### PR DESCRIPTION
**Pull request by jhalaria**
@iemejia - Please review. Thank-you.
- Noticed an issue while reading around 1000 files from S3.
- Started getting connection timeouts as the max number of open connections is set to 50.
- Inside https://github.com/apache/beam/blob/master/runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/io/UnboundedSourceWrapper.java#L250 , we open all the readers simultaneously. Thats the intended behavior for an unbounded source. But when it comes to reading a bounded source (eg. Doing reads from S3), opening all connections at the same time leads to connections not available when the number of files we are trying to open is greater than the maxHttpConnections possible which by default is set to 50 (ClientConfiguration.`DEFAULT_MAX_CONNECTIONS`) and there isn't a way to override it [We should have the ability to override this anyways. Will create a separate PR for that.]. 
- The Change essentially gives the `BoundedToUnboundedSourceAdapter` an `ArrayDeque` so that one reader can read more than 1 `BoundedSource`. Each Reader finishes reading from a `BoundedSource` and then goes to the next one.
- In case of check-pointing, if a reader has elements that aren't read yet from a `BoundedSource`, we create a checkpoint with the remaining elements of the `BoundedSource` and the elements remaining in the `ArrayDequeue`.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/UnboundedReadFromBoundedSource.java [hunk 1]
```diff
@@ -104,122 +96,102 @@ public void populateDisplayData(DisplayData.Builder builder) {
...
 
-  /** A {@code BoundedSource} to {@code UnboundedSource} adapter. */
   @VisibleForTesting
   public static class BoundedToUnboundedSourceAdapter<T>
       extends UnboundedSource<T, BoundedToUnboundedSourceAdapter.Checkpoint<T>> {
 
-    private BoundedSource<T> boundedSource;
+    //TODO: There must be a better way to figure out the maxOpenConnections available. Use that number here
+    //Mention this as part of PR to get feedback.
+    private static final int READER_QUEUE_SIZE = 10;
```
Comment (jhalaria) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/UnboundedReadFromBoundedSource.java]: @iemejia - Whats a good way to set the value here?

**Comment by jhalaria** [comment]
@iemejia - Whenever you get a chance, please review. Thanks.

**Comment by iemejia** [comment]
Ok some general comments:

- I updated the JIRA ticket to make it more clear now that we know that the issue is related to the way Flink translation works (by using runners-core `UnboundedReadFromBoundedSource`).
- For the comment on tuning `ClientConfiguration` on S3 you can already do that via [S3Options.setClientConfiguration](https://github.com/apache/beam/blob/58372ca8c183f3840fef6d6ec2fab24dd7028475/sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/options/AwsOptions.java#L104).

In general I think the solution makes sense, however I have a bit of doubt if the point where you are fixing it is the correct one, I would like to preserve the current 1-1 relation of unbounded-bounded if possible. (Note that if we can't we will need to update the current documentation because with these changes the semantics has clearly changed).

Thinking a bit this is similar to the Connection/Thread Pool classic problem and I don't know if we may create something similar where the limited 'tasks' are such readers, or if we can just delegate this into native Flink and it will take care. Since I am not a Flink expert I asked @mxm to help us take a look, maybe he knows a better way to proceed..

**Comment by jhalaria** [comment]
Thank-you @mxm and @iemejia for looking into this. I will incorporate your feedback.

**Comment by robertwb** [comment]
Any updates on this PR?

**Comment by iemejia** [comment]
Oups seems this needs a rebase too now @jhalaria

**Comment by jhalaria** [comment]
@robertwb - I was out on vacation. Haven't had a chance to resume work on it yet. Will let you know once I have something.

**Pull request closed without merging**